### ISSUE TITLE
TODO List of All Algorithms/DS to be implemented

### ISSUE THREAD

**Issue by aktech** [aktech/pydsa#12]
Binary Search
Quicksort 
Merge Sort 
Suffix Array 
Knuth-Morris-Pratt Algorithm (KMP) 
Rabin-Karp Algorithm 
Tries 
Depth First Traversal of a graph
Breadth First Traversal of a graph
Dijkstra's Algorithm
Binary Indexed Tree 
Segment Tree (with lazy propagation) 
Persistent Segment Tree,
Z algorithm 
Floyd Warshall Algorithm 
Sparse Table(RMQ)
Heap / Priority Queue / Heapsort 
Modular Multiplicative Inverse
Suffix Automaton 
Lowest Common Ancestor 
Counting Inversions
Euclid's Extended Algorithm

Suffix Tree 
Dynamic Programming: Longest Increasing Subsequence, Bitmask DP, Bitmask DP, Optimization, DP on Trees 
Basic Data Structures : Stack Implementation, Queue Implementation, Linked List Implementation
Logarithmic Exponentiation
Graphs
Minimum Spanning Tree : Kruskal's Implementation, Prim's Implementation
Efficient Prime Factorization
Combinatorics
Union Find/Disjoint Set 
Knapsack problem
Aho-Corasick String Matching Algorithm 
Strongly Connected Components
Bellman Ford algorithm 
Heavy-light Decomposition 
Convex Hull : Jarvis Algorithm Implementation,  Graham scan
Line Intersection
Sieve of Erastothenes
Interval Tree 
Counting Sort
Probabilities
Matrix Exponentiation
Network flow : (Max Flow)Tutorial : I, II, Max Flow(Ford-Fulkerson) Tutorial, Implementation, (Min Cut) Tutorial, Implementation, (Min Cost Flow)Tutorial : I, II, III, Dinic's Algorithm with Implementation, Max flow by Edmonds Karp with Implementation
K-d tree
Deque
Binary Search Tree
Quick Select
Treap/Cartesian Tree
Game Theory
Maximum Bipartite Matching
Manacher's Algorithm
Miller-Rabin Primality Test
Stable Marriage Problem
Hungarian Algorithm
Sweep line Algorithm
LCP
Gaussian Elimination
Pollard Rho Integer Factorization, problem
Topological Sorting
Detecting Cycles in a Graph
Geometry
Backtracking : N queens problem, Tug of War, Sudoku
Eulerian and Hamiltonian Paths : (Eulerian Path and Cycle)Implementation, (Hamiltonian Cycle)
Graph Coloring
Meet in the Middle 
Radix Sort, Bucket Sort
Johnson's Algorithm 
Maximal Matching in a General Graph : Blossom/Edmond's Algorithm, Implementation, Tutte Matrix
Recursion: Towers of Hanoi with explanation
Inclusion and Exclusion Principle
Co-ordinate Compression
Sqrt-Decomposition
Link-Cut Tree
Euler's Totient Function 
Burnside Lemma
Edit/Levenshtein Distance
Branch and Bound
Math for Competitive Programming
Mo's Algorithm : Tutorial and Problems

**Comment by Sumith1896** [comment]
@aktech We need classification of algos by class (`sorting`, `graphs`)

**Comment by aktech** [comment]
Yes, Indeed.

**Comment by NitikaAgarwal** [comment]
Hi, there should be some classification for the algorithms. 
### Sorting ###

- Bubble sort
- Selection Sort
- Insertion sort
- Quick sort
- Merge sort
- Counting sort
- Radix sort
- Bucket sort

### Graphs  ###
- Depth First Traversal of a graph
- Breadth First Traversal of a graph
- Dijkstra's Algorithm
- Floyd Warshall Algorithm
- Bellman Ford algorithm
- Topological Sorting
- Detecting Cycles in a Graph
- Graph Coloring

**Comment by NitikaAgarwal** [comment]
Other classification can be : 
### Binary Trees ###
### Binary Search Trees ###
### Dynamic Programming ###
### Arrays ###
### Mathematical Algorithms ###
### Heaps ###
### Stacks ###
### Queues ###
### Linked Lists ###

**Comment by souravsingh** [comment]
Can we also add Boyer-Moore to the list of Algorithms?

**Comment by aktech** [comment]
> Can we also add Boyer-Moore to the list of Algorithms?

Yes, why not. This list is not exhaustive, feel free to implement any algorithm.

**Comment by tapasweni-pathak** [comment]
Is this list updated? Can I go ahead and pick any from this list, which is unchecked and not referenced.

**Comment by CuriousLearner** [comment]
Hi @tapasweni-pathak !

I don't think this list is updated. I was reviewing several PR's which had these algos.

You need to search upon the pending tasks to know which one is not yet picked up. Thanks!

**Comment by tapasweni-pathak** [comment]
@CuriousLearner Can I reference this issue in the pull request(s) that have been created for one of these algorithm/data structure? This will be useful for the future contributors as well, to not look into every pull request to find out what they can and can't work on.

**Comment by CuriousLearner** [comment]
@tapasweni-pathak Yes, you should do it. And once your PR gets merged, mark the algo as checked here :)

**Comment by tapasweni-pathak** [comment]
@CuriousLearner I am asking for other's pull requests. For example, #23, #106 and many others are not referencing this issue anywhere in the description or in the comments of the pull request. Adding a reference to this issue there will make it easier for the new contributors, as every pull request that has code for one or more of any of these data structure/algorithm will be referencing this issue. So new contributors can simply look into all the references on this issue and start working on any which is not yet referenced.

**Comment by CuriousLearner** [comment]
@tapasweni-pathak yes, you can do, that would be great :)
